File: tools/CliShim/CommandLine.cs
// Copyright (c) Microsoft Corporation
// The Microsoft Corporation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

namespace PowerToys.CliShim;

/// <summary>
/// Helpers for working with the raw process command line. Kept in its own internal type
/// (rather than inside <see cref="Program"/>) so the parsing logic can be unit tested by
/// linking this single source file into the test project.
/// </summary>
internal static class CommandLine
{
    /// <summary>
    /// Returns the command line with its first token (argv[0]) removed, following the C
    /// runtime rule for the program name: leading whitespace is skipped first; then, when
    /// argv[0] starts with a quote it ends at the next quote (no backslash-escaping for the
    /// program name), otherwise it ends at the first whitespace. Whitespace before the first
    /// real argument is then trimmed.
    /// </summary>
    /// <param name="commandLine">The raw process command line (for example <see cref="System.Environment.CommandLine"/>).</param>
    /// <returns>The remaining arguments, verbatim.</returns>
    internal static string StripArgumentZero(string commandLine)
    {
        int index = 0;

        // Skip leading whitespace before argv[0]. The OS loader never produces this, but a
        // non-shell parent that calls CreateProcessW with a padded lpCommandLine can, and
        // without this the unquoted scan below would stall at index 0 and leak the program
        // name into the forwarded arguments.
        while (index < commandLine.Length && (commandLine[index] == ' ' || commandLine[index] == '\t'))
        {
            index++;
        }

        if (index < commandLine.Length && commandLine[index] == '"')
        {
            index++;
            while (index < commandLine.Length && commandLine[index] != '"')
            {
                index++;
            }

            if (index < commandLine.Length)
            {
                index++; // Consume the closing quote.
            }
        }
        else
        {
            while (index < commandLine.Length && commandLine[index] != ' ' && commandLine[index] != '\t')
            {
                index++;
            }
        }

        while (index < commandLine.Length && (commandLine[index] == ' ' || commandLine[index] == '\t'))
        {
            index++;
        }

        return commandLine[index..];
    }
}

File: tools/CliShim/CommandLine.cpp
// Copyright (c) Microsoft Corporation
// The Microsoft Corporation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

#include "CommandLine.h"

namespace CommandLine
{
    std::wstring StripArgumentZero(std::wstring_view commandLine)
    {
        const auto isWhitespace = [](const wchar_t character) {
            return character == L' ' || character == L'\t';
        };

        size_t index = 0;

        // A non-shell CreateProcessW caller can prepend whitespace; without this skip the scan
        // below stalls at index 0 and leaks argv[0] into the forwarded tail. This is a deliberate
        // departure from the CRT, which would report an empty argv[0] instead.
        while (index < commandLine.size() && isWhitespace(commandLine[index]))
        {
            ++index;
        }

        // argv[0] is tokenized differently from every later argument, and the rule that matters is
        // the CRT's, because that is what every target ends up parsing: FileLocksmithCLI is a
        // native wmain, and the .NET CLIs receive their string[] from the apphost's wmain. The CRT
        // (ucrt\startup\argv_parsing.cpp, parse_command_line) toggles an in-quotes flag on every
        // quote while scanning argv[0] and ends the name at the first whitespace found outside
        // quotes; a quote never terminates the name and a backslash never escapes one.
        //
        // CommandLineToArgvW is the odd one out - it ends a quoted argv[0] at the closing quote,
        // with no toggling - so following it instead would leak the rest of the program name into
        // the tail of a partially quoted command line such as
        // `"%ProgramFiles%"\PowerToys\bin\PowerToys.FancyZones.CLI.exe arg`, which cmd.exe passes
        // through verbatim. Matching the CRT is what makes the shim transparent: the target sees
        // the exact tail it would have seen had the caller invoked it directly.
        bool inQuotes = false;
        while (index < commandLine.size())
        {
            const wchar_t character = commandLine[index];
            ++index;

            if (character == L'"')
            {
                inQuotes = !inQuotes;
                continue;
            }

            if (!inQuotes && isWhitespace(character))
            {
                break;
            }
        }

        while (index < commandLine.size() && isWhitespace(commandLine[index]))
        {
            ++index;
        }

        return std::wstring{ commandLine.substr(index) };
    }
}
